### PR TITLE
[Placement Group] Atomic Creation using 2 phase protocol part 2.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

Design doc: https://docs.google.com/document/d/1tPMs2W-557hPmln5xnycK85vKTxJujZIsjq9Rsw9OzY/edit#

This is the last P0 item to implement basic placement group functionalities. It implements raylet side logic for atomic placement group creation.

Note that in order to separate the current cancel methods, I always commit resources before canceling if necessary. In this way, we can reuse the current method.

Unlike specified in the design doc, I don't guarantee the idempotency yet. It has some benefits
- Since we don't have retry yet, having idempotency has no meaning.
- Instead I added more check to guarantees invariants. This will help us detecting bugs easily if there are. I added TODO to make these methods idempotent (and it will be implemented when we implement retry).

## Related issue number

https://github.com/ray-project/ray/issues/9807

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/tests/test_placement_group.py [hunk 1]
```diff
@@ -635,48 +634,161 @@ class NormalActor:
...
-        strategy="PACK",
+        strategy="SPREAD",
         bundles=[{
             "CPU": bundle_cpu_size
         } for _ in range(num_nodes * bundle_per_node)])
 
     # Create a placement group actor.
-    # This shouldn't be scheduled until placement group creation is done.
+    # This shouldn't be scheduled because atomic
+    # placement group creation should've failed
```
Comment (ffbin) [inline python/ray/tests/test_placement_group.py:656]:
```suggestion
    # placement group creation should've failed.
```

File: src/ray/common/task/scheduling_resources.h [hunk 2]
```diff
@@ -571,16 +578,31 @@ class SchedulingResources {
...
 
   /// \brief Update total, available and load resources with the ResourceIds.
-  /// Create if not exists.
+  /// Create if not exists. This will only update resources, but it won't
+  /// create placement group resources. That'll be done when resources are
+  /// COMMITTED. Commit should be done by CommitBundleResources.
+  ///
+  /// We need this step for running 2PC protocol for atomic placement group creation.
+  ///
+  /// \param resource_name: Name of the resource to be modified
```
Comment (ffbin) [inline src/ray/common/task/scheduling_resources.h:587]:
```suggestion
  /// \param resource_name: Name of the resource to be modified.
```

File: src/ray/raylet/node_manager.h [hunk 3]
```diff
@@ -99,6 +99,28 @@ struct NodeManagerConfig {
...
+  std::size_t operator()(const std::pair<T1, T2> &pair) const {
+    return std::hash<T1>()(pair.first) ^ std::hash<T2>()(pair.second);
+  }
+};
+
+enum CommitState {
+  /// Resources are prepared.
+  PREPARE,
+  /// Resources are COMMITTED.
+  COMMIT
```
Comment (ffbin) [inline src/ray/raylet/node_manager.h:114]: /// Resources are prepared.
PREPARED,
/// Resources are COMMITTED.
COMMITTED

File: src/ray/raylet/node_manager.h [hunk 4]
```diff
@@ -99,6 +99,28 @@ struct NodeManagerConfig {
...
+};
+
+struct BundleState {
+  /// Leasing state for 2PC protocol.
+  CommitState state;
+  /// resources that are acquired at preparation stage.
```
Comment (ffbin) [inline src/ray/raylet/node_manager.h:120]:
```suggestion
  /// Resources that are acquired at preparation stage.
```

File: src/ray/raylet/node_manager.cc [hunk 5]
```diff
@@ -1824,8 +1824,17 @@ void NodeManager::HandlePrepareBundleResources(
 void NodeManager::HandleCommitBundleResources(
     const rpc::CommitBundleResourcesRequest &request,
     rpc::CommitBundleResourcesReply *reply, rpc::SendReplyCallback send_reply_callback) {
+  RAY_CHECK(!new_scheduler_enabled_) << "Not implemented yet.";
+
+  auto bundle_spec = BundleSpecification(request.bundle_spec());
+  RAY_LOG(DEBUG) << "bundle commit request " << bundle_spec.BundleId().first
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1830]:
```suggestion
  RAY_LOG(DEBUG) << "Received bundle commit request " << bundle_spec.BundleId().first
```

File: src/ray/raylet/node_manager.cc [hunk 6]
```diff
@@ -1860,6 +1869,20 @@ void NodeManager::HandleCancelResourceReserve(
     worker->MarkDead();
     KillWorker(worker);
   }
+
+  // Return resources. We should commit resources if it weren't because
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1873]: The comments and the code don't match.

Comment (rkooo567) [reply src/ray/raylet/node_manager.cc:1873]: Remove the first two words.

File: python/ray/tests/test_placement_group.py [hunk 7]
```diff
@@ -635,48 +634,161 @@ class NormalActor:
...
+
+    wait_for_condition(pg_removed)
+
+    # Make sure check without pgs are all
+    # scheduled properly because resources are cleaned up.
+    assert all(ray.get(check_without_pg))
+
+
+def test_mini_stress_scenario(ray_start_cluster):
+    # Create bundles as many as number of gpus in the cluster.
```
Comment (rkooo567) [inline python/ray/tests/test_placement_group.py:721]: This is not directly related to this PR, but I added it because we completed all P0 features.

**Review by ffbin — APPROVED** [review]
LGTM

File: src/ray/raylet/node_manager.cc [hunk 8]
```diff
@@ -1860,6 +1869,20 @@ void NodeManager::HandleCancelResourceReserve(
     worker->MarkDead();
     KillWorker(worker);
   }
+
+  // We should commit resources if it weren't because
+  // ReturnBundleResources requires resources to be committed when it is called.
+  auto it = bundle_state_map_.find(bundle_spec.BundleId());
+  RAY_CHECK(it != bundle_state_map_.end())
+      << "Cancel requests are received to raylet although it hasn't received any prepare "
+         "or commit reuqest. This must be an anomaly.";
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc:1878]:
```suggestion
         "or commit requests. This must be an anomaly.";
```

**Review by ericl — APPROVED** [review]
